core-common/src/test/java/org/zalando/nakadi/annotations/validation/DataLakeAnnotationsTest.java

```diff
@@ -0,0 +1,78 @@
+package org.zalando.nakadi.annotations.validation;
+
+import org.junit.Before;
+import org.junit.Test;
+
+import javax.validation.ConstraintViolation;
+import javax.validation.Valid;
+import javax.validation.Validation;
+import javax.validation.Validator;
+import java.util.Map;
+import java.util.Set;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class DataLakeAnnotationsTest {
+    public static class TestClass {
+        @Valid
+        @DataLakeValidAnnotations
+        private final Map<
+                @Valid @AnnotationKey String,
+                @Valid @AnnotationValue String> annotations;
+
+        public TestClass(final Map<String, String> annotations) {
+            this.annotations = annotations;
+        }
+    }
+
+    private Validator validator;
+
+    @Before
+    public void prepareValidator() {
+        validator = Validation.buildDefaultValidatorFactory().getValidator();
+    }
+
+    @Test
+    public void whenRetentionPeriodThenRetentionReasonRequired() {
+        final var annotations = Map.of(
+                DataLakeAnnotationValidator.RETENTION_PERIOD_ANNOTATION, "1 day"
+        );
+        final Set<ConstraintViolation<TestClass>> result = validator.validate(new TestClass(annotations));
+        assertFalse("Missing retention reason is treated valid", result.isEmpty());
+    }
+
+    @Test
+    public void whenRetentionPeriodFormatIsWrongThenFail() {
+        final var annotations = Map.of(
+                DataLakeAnnotationValidator.RETENTION_PERIOD_ANNOTATION, "1 airplane",
+                DataLakeAnnotationValidator.RETENTION_REASON_ANNOTATION, "I need my data"
+        );
+        final Set<ConstraintViolation<TestClass>> result = validator.validate(new TestClass(annotations));
+        assertFalse("Missing retention reason is treated valid", result.isEmpty());
```

Comment:
Copy-paste from above.  Can/should we check the actual validation result, not just that it's not empty?

Reply:
That is a good tip. I've provided better messages from the validator and extra use cases as well.